File: app/src/main/java/com/przemolab/oknotifier/sync/RetrieveContestsTask.java
package com.przemolab.oknotifier.sync;

import android.os.AsyncTask;

import com.przemolab.oknotifier.fragments.ContestsListFragment;
import com.przemolab.oknotifier.models.Contest;
import com.przemolab.oknotifier.modules.NotifierRepository;
import com.przemolab.oknotifier.modules.OpenKattisService;

import java.util.List;

import timber.log.Timber;

public class RetrieveContestsTask extends AsyncTask<Void, Void, List<Contest>> {

    private OpenKattisService openKattisService;
    private NotifierRepository notifierRepository;
    private ContestsListFragment.OnContestsListEventsListener onContestsListEventsListener;

    public RetrieveContestsTask(OpenKattisService openKattisService, NotifierRepository notifierRepository,
                               ContestsListFragment.OnContestsListEventsListener onContestsListEventsListener) {
        this.openKattisService = openKattisService;
        this.notifierRepository = notifierRepository;
        this.onContestsListEventsListener = onContestsListEventsListener;
    }

    @Override
    protected void onPreExecute() {
        onContestsListEventsListener.onContestSyncStarted();
    }

    @Override
    protected List<Contest> doInBackground(Void...voids) {
        try {
            List<Contest> contests = openKattisService.getOngoingContests();
            notifierRepository.persistContests(contests);
            return contests;
        } catch (Exception ex) {
            Timber.e(ex);
        }
        return null;
    }

    @Override
    protected void onPostExecute(List<Contest> contests) {
        onContestsListEventsListener.onContestSyncFinished(contests);
    }
}

File: app/src/main/java/com/przemolab/oknotifier/sync/RetrieveContestsTask.kt
package com.przemolab.oknotifier.sync

import android.os.AsyncTask
import com.przemolab.oknotifier.data.entries.ContestEntry

import com.przemolab.oknotifier.fragments.ContestsListFragment
import com.przemolab.oknotifier.interfaces.INotifierRepository
import com.przemolab.oknotifier.interfaces.IOpenKattisService

import timber.log.Timber

class RetrieveContestsTask(private val openKattisService: IOpenKattisService,
                           private val notifierRepository: INotifierRepository,
                           private val onContestsListEventsListener: ContestsListFragment.OnContestsListEventsListener) : AsyncTask<Void, Void, List<ContestEntry>>() {

    override fun onPreExecute() {
        onContestsListEventsListener.onContestSyncStarted()
    }

    override fun doInBackground(vararg voids: Void): List<ContestEntry>? {
        try {
            val contests = openKattisService.ongoingContests
            notifierRepository.persistContests(contests)
            return contests
        } catch (ex: Exception) {
            Timber.e(ex)
        }

        return null
    }

    override fun onPostExecute(contestEntries: List<ContestEntry>) {
        onContestsListEventsListener.onContestSyncFinished(contestEntries)
    }
}
